lib/ansible/modules/windows/win_copy.ps1

```diff
@@ -1,94 +1,141 @@
     changed = $false
-    dest = $dest
-    original_basename = $original_basename
-    src = $src
 }
 
-if (($force -eq $false) -and (Test-Path -Path $dest)) {
-    $result.msg = "file already exists"
-    Exit-Json $result
+if ($diff_mode) {
+    $result.diff = @{}
 }
 
-Function Copy-Folder($src, $dest) {
-    if (Test-Path -Path $dest) {
-        if (-not (Get-Item -Path $dest -Force).PSIsContainer) {
-            Fail-Json $result "If src is a folder, dest must also be a folder. src: $src, dest: $dest"
+Function Copy-File($source, $dest) {
+    $diff = ""
+    $copy_file = $false
+    $source_checksum = $null
+    if ($force) {
+        $source_checksum = Get-FileChecksum -path $source
+    }
+
+    if (Test-Path -Path $dest -PathType Container) {
+        Fail-Json -obj $result -message "cannot copy file from $source to $($dest): dest is already a folder"
+    } elseif (Test-Path -Path $dest -PathType Leaf) {
+        if ($force) {
+            $target_checksum = Get-FileChecksum -path $dest
+            if ($source_checksum -ne $target_checksum) {
+                $copy_file = $true
+            }
         }
     } else {
-        try {
-            New-Item -Path $dest -ItemType Directory -Force -WhatIf:$check_mode
-            $result.changed = $true
-        } catch {
-            Fail-Json $result "Failed to create new folder $dest $($_.Exception.Message)"
-        }
+        $copy_file = $true
     }
 
-    foreach ($item in Get-ChildItem -Path $src) {
-        $dest_path = Join-Path -Path $dest -ChildPath $item.PSChildName
-        if ($item.PSIsContainer) {
-            Copy-Folder -src $item.FullName -dest $dest_path
-        } else {
-            Copy-File -src $item.FullName -dest $dest_path
+    if ($copy_file) {
+        $file_dir = [System.IO.Path]::GetDirectoryName($dest)
+        # validate the parent dir is not a file and that it exists
+        if (Test-Path -Path $file_dir -PathType Leaf) {
+            Fail-Json -obj $result -message "cannot copy file from $source to $($dest): object at dest parent dir is not a folder"
+        } elseif (-not (Test-Path -Path $file_dir)) {
+            # directory doesn't exist, need to create
+            New-Item -Path $file_dir -ItemType Directory -WhatIf:$check_mode | Out-Null
+            $diff += "+$file_dir`n"
         }
-    }
-}
 
-Function Copy-File($src, $dest) {
-    if (Test-Path -Path $dest) {
-        if ((Get-Item -Path $dest -Force).PSIsContainer) {
-            Fail-Json $result "If src is a file, dest must also be a file. src: $src, dest: $dest"
+        if (Test-Path -Path $dest -PathType Leaf) {
+            Remove-Item -Path $dest -Force -Recurse | Out-Null
+            $diff += "-$dest`n"
+        }
+
+        if (-not $check_mode) {
+            # cannot run with -WhatIf:$check_mode as if the parent dir didn't
+            # exist and was created above would still not exist in check mode
+            Copy-Item -Path $source -Destination $dest -Force | Out-Null
```

Comment:
And this shouldn't be a concern if the module doesn't create the directory. (Obviously this module would still fail if the directory was created in the same playbook run in check-mode, but also not our concern)